src/computability/regex.lean

```diff
@@ -0,0 +1,216 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Fox Thomson.
+-/
+import data.fintype.basic
+import data.finset.basic
+import tactic.rcases
+import tactic.omega
+
+/-!
+# Regular Expression
+
+This file contains the formal definition for regular expressions on basic lemmas.
+-/
+
+universe u
+
+variables {α : Type u} [dec : decidable_eq α]
+
+/-- This is the definition of regular expressions
+  `RZero` matches nothing
+  `RNull` matches only the empty string
+  `RChar a` matches only the string 'a'
+  `RStar M` matches any finite concatenation of strings which match `M`
+  `RPlus M N` matches anything which match `M` or `N`
```

Comment:
I'd expect `RPlus` to describe a `+` like in the regex `a+`, but it seems you're using it to mean the `|` in `a|b`. Would `ROr` be a better name?

Reply:
I chose `RPlus` to reflect the definition of a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra) but I see the confusion. I think this may be part of a bigger problem in that CS regexes like the IEEE POSIX standard allows for things like backreferences which make them non-regular (and also non-context free) languages. Maybe the file name should be changed to `regular_expression` and reserve regex for talking about things with backrefs, etc (see https://en.wikipedia.org/wiki/Regular_expression#Patterns_for_non-regular_languages)

Reply:
Phrased another way, the problem is that I have a CS backgruond of thinking of the posix standard :). Putting something in the docstring making clear that this is _not_ the normal CS regex would certainly help, whether or not you rename.